### ISSUE TITLE
Build of 1.4 fail with mingw32 in win10 and curl 7.74.0

### ISSUE THREAD

**Issue by terablade2001** [JosephP91/curlcpp#131]
Hello.
I tried to compile the library (latest master branch and tag1.4) in Windows 10 with mingw32 and (latest) curl 7.74.0 (x32 and x64). The curl is downloaded from https://curl.se/windows/.

The static library libcurlcpp.a is created with no problem.

But when it tries to link with `recv_header_example_exec.exe` (or other programs) it get undefined reference errors.
I've included the compile procedure's log below for commit-tag 1.4 with curl
[compilex32URFail.log](https://github.com/JosephP91/curlcpp/files/5681691/compilex32URFail.log)

When trying to link with curl 7.74.0 x64 I get similar issues. Same when linking with the `libcurl.dll.a`.

If anyone could check it it would be nice.
Thanks.

**Comment by JosephP91** [comment]
Hello @terablade2001 

can you retry the cmake command adding these options? 

`-DCURL_STATICLIB -static -lstdc++ -lgcc -lpthread -lcurl -lwldap32 -lws2_32`

**Comment by terablade2001** [comment]
Hi. I just tried:
`cmake -G "MinGW Makefiles" -DCMAKE_BUILD_TYPE=Release -DCURL_LIBRARY=c:/git/curl-7.74.0-win32-mingw/lib/libcurl.a -DCURL_INCLUDE_DIR=c:/git/curl-7.74.0-win32-mingw/include -DCURL_STATICLIB=True -static -lstdc++ -lgcc -lpthread -lcurl -lwldap32 -lws2_32 ..`
Problem seems to still remain.
[compilex32URFail2.log](https://github.com/JosephP91/curlcpp/files/5682862/compilex32URFail2.log)

**Comment by JosephP91** [comment]
Can you try to link agasint libcurl.dll.a instead of libcurl.a?

**Comment by terablade2001** [comment]
Attached the results below
[compilex32URFail3.log](https://github.com/JosephP91/curlcpp/files/5682900/compilex32URFail3.log)
It's a similar behavior.

**Comment by JosephP91** [comment]
definitely it's a libcurl linking problem. Last try: when you link against the libcurl.dll.a, remove the -`DCURL_STATICLIB=True -static` flags. If it doesn't work, I have to dig more into it, because I am not on a Windows machine.

**Comment by terablade2001** [comment]
It didn't worked either.
[compilex32URFail4.log](https://github.com/JosephP91/curlcpp/files/5682914/compilex32URFail4.log)

Now that you are mention it, maybe indeed the dll is not compatible with mingw.
The link of curl (https://curl.se/windows/) has a log saying:
> 
>  The following tools/compilers were used in the build process:
>     binutils-mingw-w64-i686 2.35
>     binutils-mingw-w64-x86_64 2.35
>     clang 11.0.0
>     gcc-mingw-w64-i686 10-win32
>     gcc-mingw-w64-x86_64 10-win32
>     mingw-w64 8.0.0-1 

I'm using TDM-GCC compiler from here.
https://jmeubank.github.io/tdm-gcc/

If you have any time to check this out, it would be useful.
Thanks anyway.

**Comment by JosephP91** [comment]
Hi @terablade2001 can I ask you which cmake and make are you using? As far as I can see, the TDM-GCC installation does not install cmake or make. Am I right?

**Comment by terablade2001** [comment]
Hi.
TDM-GCC has it's own make. For cmake I use an external package. Below is my setup with more info.
Just to note that I've copy `mingw32-make.exe` to `make.exe` in the same folder `C:\TDM-GCC-64\bin\` (_for my convenience_):
```C:\Windows\system32>where cmake
C:\Program Files\CMake\bin\cmake.exe

C:\Windows\system32>cmake -version
cmake version 3.18.0-rc1

CMake suite maintained and supported by Kitware (kitware.com/cmake).

C:\Windows\system32>where make
C:\TDM-GCC-64\bin\make.exe

C:\Windows\system32>make --version
GNU Make 3.82.90
Built for i686-pc-mingw32
Copyright (C) 1988-2012 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.```

**Comment by JosephP91** [comment]
@terablade2001 can you try to add `curl` in the target_link_libraries instructions in the `test/CMakeLists.txt` ? 

You should have this situation after that:

```
target_link_libraries(cookie_example_exec curlcpp curl)
target_link_libraries(custom_request_example_exec curlcpp curl)
target_link_libraries(easy_example_exec curlcpp curl)
target_link_libraries(easy_info_example_exec curlcpp curl)
target_link_libraries(header_example_exec curlcpp curl)
target_link_libraries(recv_header_example_exec curlcpp curl)
```

**Comment by JosephP91** [comment]
Regarding the multiple definitions, I don't know where they're coming from. I don't have those errors on linux

**Comment by terablade2001** [comment]
Indeed using the curl in target_link_libraries() solves the undefined behavior.

I did experimented a bit with your code, and I can reduce redefinition errors if I place the non-inline methods from header files to cpp files, leaving in the hpp files only the declarations of the methods.
However I think there is a problem with mingw or the linker.
The most similar result I could find, describing the behavior I see, is described at the link
https://sourceware.org/bugzilla/show_bug.cgi?id=12762
where if the same header file, containing definitions/implementations, is used on more than one object files (_i.e. to compile curlcpp, and then to compile the test example_), linker fails to recognize the same symbols. By default -flto seems to enabled, but if I disable it I have to mess up with plugins.

The only solution (which is not trustworthy, but may work for my needs) is to use `-Wl,-allow-multiple-definition` in the linker.

I also had to add the `link_directories("c:/git/curl-7.74.0-win64-mingw/lib/")` in the first line of `test/CMakeLists`. [Version 64, not 32].
My cmake now is:
`cmake -G "MinGW Makefiles" -DCMAKE_BUILD_TYPE=Release -DCURL_LIBRARY=c:/git/curl-7.74.0-win64-mingw/lib/libcurl.dll.a -DCURL_INCLUDE_DIR=c:/git/curl-7.74.0-win64-mingw/include -DCMAKE_EXE_LINKER_FLAGS="-Wl,-allow-multiple-definition" ..`

Then it successfully compiles and links the first two targets, but it fails to include "curl_easy.h" in the third target (!!!!)
[compilex32URFail6.log](https://github.com/JosephP91/curlcpp/files/5691471/compilex32URFail6.log)

Just to note, testing the created .exe files, it seems they require also other .dlls (libcrypt, libssl, etc) which are not available to my downloaded curl packages (!). I guess I'll have to go in linux/ubuntu to do proper work...

Anyway. Thanks a lot for your time! :)
